### PR TITLE
Send rows in binary mode for ANALYZE

### PR DESCRIPTION
**Pull request by seryozhasmirnov**
Send results of `select pg_catalog.gp_acquire_sample_rows`  query in
binary mode.
That allows to avoid overflow for max double.

For example, if run the following prior to this fix:
```
set extra_float_digits to 0;
create table t (a double precision);
insert into t values (1.7976931348623157e+308);
analyze t;
```

the following message will be printed:

`ERROR:  value out of range: overflow`

For text mode (default) when analyze for table is performed the master
segment calls `gp_acquire_sample_rows()` helper function on each
segment. That eventually calls float8out function on segment to
converts `float8` number to a string with `snprintf`:

`int	ndig = DBL_DIG + extra_float_digits;
`
`snprintf(ascii, MAXDOUBLEWIDTH + 1, "%.*g", ndig, num);
`
When ndig is 15 the maximum float8 value **1.7976931348623157e+308** is
rounded to "**1.79769313486232e+308**" that has no representation.

And on master `acquire_sample_rows_dispatcher` function
process `gp_acquire_sample_rows` result and eventually `float8in`
function is called to convert string to `float8` with `strtold`:
`val = strtold(num, &endptr);
`
This is where overflow for "**1.79769313486232e+308**" happens but works
fine for "**1.7976931348623157e+308**".

Transferring in binary mode allows to avoid conversion from double to
string on segments and then back to double on master.

Using `CdbDispatchPlan` instead of `CdbDispatchCommand` allows
to receive data in binary mode in `MemTuple`.

As `acquire_sample_rows_dispatcher` executes
`select pg_catalog.gp_acquire_sample_rows` with `CdbDispatchPlan`
when `Test_print_direct_dispatch_info` is enabled `cdbdisp_dispatchX`
prints `INFO message  "(slice %d) Dispatch command to %s"`.
This has to be added to expected tests output.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by ashwinstar** [comment]
In-general reads this should help performance as well, apart from fixing the functional piece. Did you evaluate the code change from that front?

**Comment by ashwinstar** [comment]
Also, there are failing tests with the patch..
```
========================
 5 of 595 tests failed. 
========================


distributed_transactions
gp_copy_dtx
analyze
gporca
qp_targeted_dispatch
```

**Comment by seryozhasmirnov** [comment]
> Also, there are failing tests with the patch..
> 
> ```
> ========================
>  5 of 595 tests failed. 
> ========================
> 
> 
> distributed_transactions
> gp_copy_dtx
> analyze
> gporca
> qp_targeted_dispatch
> ```

Thanks for noting. I am working on the tests now.

**Comment by seryozhasmirnov** [comment]
> Also, there are failing tests with the patch..
> 
> ```
> ========================
>  5 of 595 tests failed. 
> ========================
> 
> 
> distributed_transactions
> gp_copy_dtx
> analyze
> gporca
> qp_targeted_dispatch
> ```

Dear @ashwinstar, I have fixed some tests and now working with analyze test fail.

I see that if tailed due to different `pg_stats.correlation` value. I am wondering why there is such test as [according to docs](https://gpdb.docs.pivotal.io/6-17/admin_guide/intro/about_statistics.html) _Greenplum Database does not calculate the correlation statistic_?
Can anyone clarify it?
Thank you!

**Comment by ashwinstar** [comment]
I think you are right, someone from team will reply to you on statistics question.
Though I am still curious to know, why changing the way we send the rows binary or non-binary, have affect to these tests?

**Comment by seryozhasmirnov** [comment]
> In-general reads this should help performance as well, apart from fixing the functional piece. Did you evaluate the code change from that front?

The following simple test shows that now analyze completes faster:
```
\timing
CREATE TABLE FLOATS(a float8);
INSERT INTO FLOATS select i FROM generate_series(1, 10000, 0.1) i;
ANALYZE floats;
```



Without changes:

> postgres=# CREATE TABLE FLOATS(a float8);
> NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
> HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
> CREATE TABLE
> Time: 10.074 ms
> postgres=# INSERT INTO FLOATS select i FROM generate_series(1, 10000, 0.1) i;
> INSERT 0 99991
> Time: 166.840 ms
> postgres=# ANALYZE floats;
> ANALYZE
> Time: **117.535 ms**
> postgres=# 

With this fix:

> postgres=# CREATE TABLE FLOATS(a float8);
> NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
> HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
> CREATE TABLE
> Time: 9.230 ms
> postgres=# INSERT INTO FLOATS select i FROM generate_series(1, 10000, 0.1) i;
> INSERT 0 99991
> Time: 213.609 ms
> postgres=# ANALYZE floats;
> ANALYZE
> Time: **96.117 ms**
> postgres=# 


The performance was not the reason of this change so improvements in performance is just a nice side effect.

**Comment by seryozhasmirnov** [comment]
> I think you are right, someone from team will reply to you on statistics question.
> Though I am still curious to know, why changing the way we send the rows binary or non-binary, have affect to these tests?

The test was affected because of the order of sample rows received from segments in response to ` select pg_catalog.gp_acquire_sample_rows`  that is used by ANALYZE..
I have added the tracing to print rows[sampleTuples] in acquire_sample_rows_dispatcher both for master ([there](https://github.com/greenplum-db/gpdb/blob/master/src/backend/commands/analyze.c#L2388-L2389)) and for code in this PR ([there](https://github.com/greenplum-db/gpdb/blob/618633b8963edd05c37855825c7c340264cdb1af/src/backend/commands/analyze.c#L2353-L2356)).

As a result, for [the following test](https://github.com/greenplum-db/gpdb/blob/618633b8963edd05c37855825c7c340264cdb1af/src/test/regress/sql/gpsd.sql#L20-L26):
```
create table gpsd_foo(a int, s text) partition by range(a);
create table gpsd_foo_1 partition of gpsd_foo for values from (1) to (5);
insert into gpsd_foo values(1, 'something');
insert into gpsd_foo values(2, chr(1000));
insert into gpsd_foo values(3, chr(105));
insert into gpsd_foo values(4, 'a \ and a "');
analyze gpsd_foo;
```

Analyze prints rows in different order:
For **master** branch:
```
gpsd_db_without_hll=# analyze gpsd_foo;
NOTICE:  rows[00]:  2  Ϩ
NOTICE:  rows[01]:  3  i
NOTICE:  rows[02]:  4  a \ and a "
NOTICE:  rows[03]:  1  something
ANALYZE
gpsd_db_without_hll=# 
```

For this PR:
```
gpsd_db_without_hll=# analyze gpsd_foo;
NOTICE:  rows[00]:  1  something
NOTICE:  rows[01]:  2  Ϩ
NOTICE:  rows[02]:  3  i
NOTICE:  rows[03]:  4  a \ and a "
ANALYZE
gpsd_db_without_hll=# 
```

The order of sample rows affects correlation statistics.

And just for information - data distribution across segments:

For **master** branch:

```
gpsd_db_without_hll=# select gp_segment_id, * from gpsd_foo;
 gp_segment_id | a |      s      
---------------+---+-------------
             0 | 2 | Ϩ
             0 | 3 | i
             0 | 4 | a \ and a "
             1 | 1 | something
(4 rows)

gpsd_db_without_hll=# 
```
For this PR:
```
gpsd_db_without_hll=# select gp_segment_id, * from gpsd_foo;
 gp_segment_id | a |      s      
---------------+---+-------------
             1 | 1 | something
             0 | 2 | Ϩ
             0 | 3 | i
             0 | 4 | a \ and a "
(4 rows)

gpsd_db_without_hll=# 
```

It shows that data is distributes across segments in the same way, but for master in analyze the data from segment id = 1 comes before data from segment with id 0, but in this PR it happens otherwise - from 0 and then from 1.

The same for [analyze test](https://github.com/greenplum-db/gpdb/blob/master/src/test/regress/sql/analyze.sql#L483-L495).

The result the query` SELECT * FROM pg_stats WHERE tablename = 'ana_parent'; ` output does not match in correlation values:
This is from master branch:
```
postgres=# select correlation FROM pg_stats WHERE tablename = 'ana_parent';
 correlation 
-------------
  0.38961038
(1 row)

postgres=#
```

This is with this PR:

```
postgres=# select correlation FROM pg_stats WHERE tablename = 'ana_parent';
 correlation 
-------------
  0.24805194
(1 row)

postgres=#
```

This is also happens due to different order of tuples received from segments:

This is for **master** branch:
```
postgres=# ANALYZE verbose ana_parent;
INFO:  analyzing "public.ana_parent"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(16384, 10000, 'f');
INFO:  analyzing "public.ana_parent" inheritance tree
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(16384, 10000, 't');
NOTICE:  rows[00]:  2
NOTICE:  rows[01]:  3
NOTICE:  rows[02]:  4
NOTICE:  rows[03]:  7
NOTICE:  rows[04]:  8
NOTICE:  rows[05]: 16
NOTICE:  rows[06]: 18
NOTICE:  rows[07]: 19
NOTICE:  rows[08]:  1
NOTICE:  rows[09]: 12
NOTICE:  rows[10]: 15
NOTICE:  rows[11]: 20
NOTICE:  rows[12]:  5
NOTICE:  rows[13]:  6
NOTICE:  rows[14]:  9
NOTICE:  rows[15]: 10
NOTICE:  rows[16]: 10
NOTICE:  rows[17]: 11
NOTICE:  rows[18]: 13
NOTICE:  rows[19]: 14
NOTICE:  rows[20]: 17
ANALYZE
postgres=# 
```
This is for this PR:
```
postgres=# ANALYZE verbose ana_parent;
INFO:  analyzing "public.ana_parent"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(16384, 10000, 'f');
INFO:  analyzing "public.ana_parent" inheritance tree
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(16384, 10000, 't');
NOTICE:  rows[00]:  5
NOTICE:  rows[01]:  6
NOTICE:  rows[02]:  9
NOTICE:  rows[03]: 10
NOTICE:  rows[04]: 10
NOTICE:  rows[05]: 11
NOTICE:  rows[06]: 13
NOTICE:  rows[07]: 14
NOTICE:  rows[08]: 17
NOTICE:  rows[09]:  1
NOTICE:  rows[10]: 12
NOTICE:  rows[11]: 15
NOTICE:  rows[12]: 20
NOTICE:  rows[13]:  2
NOTICE:  rows[14]:  3
NOTICE:  rows[15]:  4
NOTICE:  rows[16]:  7
NOTICE:  rows[17]:  8
NOTICE:  rows[18]: 16
NOTICE:  rows[19]: 18
NOTICE:  rows[20]: 19
ANALYZE
postgres=#
```

The table `ana_parent` distribution by segments as follows:
For **master** branch:
```
postgres=# select gp_segment_id, * from ana_parent;
 gp_segment_id | aa 
---------------+----
             0 |  2
             0 |  3
             0 |  4
             0 |  7
             0 |  8
             0 | 16
             0 | 18
             0 | 19
             1 |  1
             1 | 12
             1 | 15
             1 | 20
             2 |  5
             2 |  6
             2 |  9
             2 | 10
             2 | 10
             2 | 11
             2 | 13
             2 | 14
             2 | 17
(21 rows)

postgres=#
```

For this PR:
```
postgres=# select gp_segment_id, * from ana_parent;
 gp_segment_id | aa 
---------------+----
             1 |  1
             1 | 12
             1 | 15
             1 | 20
             0 |  2
             0 |  3
             0 |  4
             0 |  7
             0 |  8
             0 | 16
             0 | 18
             0 | 19
             2 |  5
             2 |  6
             2 |  9
             2 | 10
             2 | 10
             2 | 11
             2 | 13
             2 | 14
             2 | 17
(21 rows)

postgres=#
```

This is the same despite the different order of tuples received (from segments 0, 1, 2 for master and 1, 0, 2 for this PR).

But [as I pointed before](https://github.com/greenplum-db/gpdb/pull/12607#issuecomment-926829021), it is probably wrong to test `pg_stats.correlation` value when [according to docs](https://gpdb.docs.pivotal.io/6-17/admin_guide/intro/about_statistics.html) _Greenplum Database does not calculate the correlation statistic_.

**Comment by khannaekta** [comment]
w/ @reductionista 
We took a look at the `pg_stats.correlation` value in pg_stats and for GPDB, the correlation calculation done as part of analyze is used for cost estimates for indexscans in planner ([link to code](https://github.com/greenplum-db/gpdb/blob/master/src/backend/utils/adt/selfuncs.c#L6210-L6226))
The commit (f995c59) that added computing correlation was to align the analyze code with Upstream Postgres, prior to which correlation was never set. We should probably update the docs to reflect the same.

Previously, GPDB used [acquire_sample_rows_by_query()](https://github.com/greenplum-db/gpdb/blob/25de0b879d3f2b44949ab2772c1f6d6ca2fc8531/src/backend/commands/analyze.c#L358-L359), which would return undefined order of tuples, thus not calculating correlation. But with `acquire_sample_rows_dispatcher()` => calls `gp_acquire_sample_rows()`(using block sampling) on each segment and then merge the results, we can capture the correlation between the physical order of table tuples and the ordering of data values.
> As sample rows in analyze can arrive from segments in undefined order - what is the point of calculating correlation?
> > A "correlation" slot describes the correlation between the physical order of table tuples and the ordering of data values of this column.
> Probably this is the reason why it is said that Greenplum Database does not calculate the correlation statistic.

File: src/backend/commands/analyze.c [hunk 1]
```diff
@@ -2235,15 +2100,77 @@ acquire_sample_rows_dispatcher(Relation onerel, bool inh, int elevel,
...
+								destReceiver,
+								NULL,
+								NULL,
+								INSTRUMENT_NONE);
+
+	/* Call ExecutorStart to prepare the plan for execution */
+	ExecutorStart(queryDesc, 0);
+
+	/* Run the plan  */
+	ExecutorRun(queryDesc, ForwardScanDirection, 0L, true);
```
Comment (khannaekta) [inline src/backend/commands/analyze.c:2155]: This probably doesn't guarantee, the order returned for the sample data, which maybe the reason for the failing correlation tests

**Comment by seryozhasmirnov** [comment]
> w/ @reductionista We took a look at the `pg_stats.correlation` value in pg_stats and for GPDB, the correlation calculation done as part of analyze is used for cost estimates for indexscans in planner ([link to code](https://github.com/greenplum-db/gpdb/blob/master/src/backend/utils/adt/selfuncs.c#L6210-L6226)) The commit ([f995c59](https://github.com/greenplum-db/gpdb/commit/f995c59f349917a99407e3811d9b317d5d77b257)) that added computing correlation was to align the analyze code with Upstream Postgres, prior to which correlation was never set. We should probably update the docs to reflect the same.
> 
> Previously, GPDB used [acquire_sample_rows_by_query()](https://github.com/greenplum-db/gpdb/blob/25de0b879d3f2b44949ab2772c1f6d6ca2fc8531/src/backend/commands/analyze.c#L358-L359), which would return undefined order of tuples, thus not calculating correlation. But with `acquire_sample_rows_dispatcher()` => calls `gp_acquire_sample_rows()`(using block sampling) on each segment and then merge the results, we can capture the correlation between the physical order of table tuples and the ordering of data values.
> 
> > As sample rows in analyze can arrive from segments in undefined order - what is the point of calculating correlation?
> > > A "correlation" slot describes the correlation between the physical order of table tuples and the ordering of data values of this column.
> > > Probably this is the reason why it is said that Greenplum Database does not calculate the correlation statistic.

Thank you for reply, I will check the code.
But let me clarify what I wanted to ask.
In the example above I showed that tuples are received from segments with id 0, then 1 then 2 for master code and with id 2, 1, 0 for this PR (the order can be different). There order of tuples is the same, the order of segments differs.
What is the meaning of the correlation when there are several segments? Maybe it is better to calculate correlation for each segment - at least there ordering will not be impacted by data from other segments.
Correlation depends on the order of sample rows received from segment. If one segment is faster then others at one run, but slower in another - the correlation values will differ - that why the test has failed.
Why there is a such dependency and what is the use of it?

Thank you!

File: src/backend/commands/analyze.c (hunk 1 above)
Comment (seryozhasmirnov) [reply src/backend/commands/analyze.c:2155]: > This probably doesn't guarantee, the order returned for the sample data, which maybe the reason for the failing correlation tests

Dear @khannaekta,

Yes, the order in what segments returns sample data impacts the correlation statistics [as in my earlier comment](https://github.com/greenplum-db/gpdb/pull/12607#issuecomment-929072188).
But why there is a such dependency? It is a parallel data processing - why correlation value should depend on order segments reply with sample data? As each segment has its own data should not there be a correlation value for each segment? 

Thank you!

Comment (khannaekta) [reply src/backend/commands/analyze.c:2155]: I don't have the complete context on calculation for correlation, so I'd recommend posting it to the community at gpdb-dev@greenplum.org

**Comment by ashwinstar** [comment]
@zxuejing I see you opened PR https://github.com/greenplum-db/gpdb/pull/15159 to take the work in this PR forward. Please can you comment/respond to the question which was raised in this PR related to Correlation.

**Comment by kainwen** [comment]
> @zxuejing I see you opened PR https://github.com/greenplum-db/gpdb/pull/15159 to take the work in this PR forward. Please can you comment/respond to the question which was raised in this PR related to Correlation.

@ashwinstar there are several improvement and fix and next steps. will post to gpdev and new pr lage soon

**Comment by zxuejing** [comment]
> @zxuejing I see you opened PR #15159 to take the work in this PR forward. Please can you comment/respond to the question which was raised in this PR related to Correlation.

hi @ashwinstar,  I have commented the question related to Correlation in the comments of new pr#15159, please check.
And i will close this pr.

**Pull request closed without merging**